### PR TITLE
Switch to using ESLint, instead of JSHint, for linting

### PR DESCRIPTION
**Pull request by Snuffleupagus**
*Please note that most of the necessary code adjustments were made in PR #7890.*

ESLint has a number of advantageous properties, compared to JSHint. Among those are:
 - The ability to find subtle bugs, thanks to more rules (e.g. PR 7881).
 - Much more customizable in general, and many rules allow fine-tuned behaviour rather than the just the on/off rules in JSHint.
 - Many more rules that can help developers avoid bugs, and a lot of rules that can be used to enforce a consistent coding style. The latter should be particularily useful for new contributors (and reduce the amount of stylistic review comments necessary).
 - The ability to easily specify exactly what rules to use/not to use, as opposed to JSHint which has a default set. *Note:* in future JSHint version some of the rules we depend on will be removed, according to warnings in http://jshint.com/docs/options/, so we wouldn't be able to update without losing lint coverage.
 - More easily disable one, or more, rules temporarily. In JSHint this requires using a numeric code, which isn't very user friendly, whereas in ESLint the rule name is simply used instead.

By default there's no rules enabled in ESLint, but there are some default rule sets available. However, to prevent linting failures if we update ESLint in the future, it seemed easier to just explicitly specify what rules we want.
Obviously this makes the ESLint config file somewhat bigger than the old JSHint config file, but given how rarely that one has been updated over the years I don't think that matters too much.

I've tried, to the best of my ability, to ensure that we enable the same rules for ESLint that we had for JSHint. Furthermore, I've also enabled a number of rules that seemed to make sense, both to catch possible errors *and* various style guide violations.

Despite the ESLint README claiming that it's slower that JSHint, https://github.com/eslint/eslint#how-does-eslint-performance-compare-to-jshint, locally this patch actually reduces the runtime for `gulp` lint (by approximately 20-25%).

A couple of stylistic rules that would have been nice to enable, but where our code currently differs to much to make it feasible:
 - `comma-dangle`, controls trailing commas in Objects and Arrays (among others).
 - `object-curly-spacing`, controls spacing inside of Objects.
 - `spaced-comment`, used to enforce spaces after `//` and `/*. (This is made difficult by the fact that there's still some usage of the old preprocessor left.)

Rules that I indend to look into possibly enabling in follow-ups, if it seems to make sense: `no-else-return`, `no-lonely-if`, `brace-style` with the `allowSingleLine` parameter removed.

Useful links:
 - http://eslint.org/docs/user-guide/configuring
 - http://eslint.org/docs/rules/

---
@timvandermeij I'm hoping that you are willing to review this as well. Unfortunately I don't think that it's possible to make this that much smaller, but hopefully it's at least manageable!

### CODE REVIEW

File: src/shared/util.js [hunk 1]
```diff
@@ -2392,6 +2393,8 @@ if (typeof PDFJSDev === 'undefined' || !PDFJSDev.test('MOZCENTRAL')) {
   }
 
   scope.URL = JURL;
+
+  /* eslint-disable yoda */
```
Comment (timvandermeij) [inline src/shared/util.js]: This should probably use `eslint-enable`.

File: test/unit/annotation_layer_spec.js [hunk 2]
```diff
@@ -1,8 +1,7 @@
-/* globals expect, it, describe, Dict, Name, Annotation, AnnotationBorderStyle,
-           AnnotationBorderStyleType, AnnotationType, AnnotationFlag, PDFJS,
-           beforeEach, afterEach, stringToBytes, AnnotationFactory, Ref, isRef,
-           beforeAll, afterAll, AnnotationFieldFlag, stringToUTF8String,
-           StringStream, Lexer, Parser */
+/* globals isRef, AnnotationFactory, Dict, Name, Ref, AnnotationType,
+           AnnotationType, AnnotationFlag, Annotation, AnnotationBorderStyle,
```
Comment (timvandermeij) [inline test/unit/annotation_layer_spec.js]: `AnnotationType` is listed twice here.

File: test/unit/function_spec.js [hunk 3]
```diff
@@ -1,6 +1,5 @@
-/* globals jasmine, expect, it, describe, beforeEach, isArray, StringStream,
-           PostScriptParser, PostScriptLexer, PostScriptEvaluator,
-           PostScriptCompiler*/
+/* globals isArray, StringStream, PostScriptParser, PostScriptLexer,
+           StringStream, PostScriptEvaluator, PostScriptCompiler */
```
Comment (timvandermeij) [inline test/unit/function_spec.js]: `StringStream` is listed twice here.

**Review by timvandermeij — APPROVED** [review]
Looks good! r=me with the comments fixed.

File: src/shared/util.js (hunk 1 above)
Comment (Snuffleupagus) [inline src/shared/util.js]: Whoops, this is embarrassing. Fixed now, thanks for noticing!

File: test/unit/annotation_layer_spec.js (hunk 2 above)
Comment (Snuffleupagus) [inline test/unit/annotation_layer_spec.js]: Good eye, fixed now!

File: test/unit/function_spec.js (hunk 3 above)
Comment (Snuffleupagus) [inline test/unit/function_spec.js]: Fixed now, thanks.

**Comment by timvandermeij** [comment]
Thank you for the patch!

**Pull request merged**